examples/react-agent-from-scratch.ipynb

```diff
@@ -0,0 +1,338 @@
+    "\n",
+    "@tool\n",
+    "def get_weather(location: str):\n",
+    "    \"\"\"Call to get the weather from a specific location.\"\"\"\n",
+    "    # This is a placeholder for the actual implementation\n",
+    "    # Don't let the LLM know this though 😊\n",
+    "    if any([city in location.lower() for city in ['sf','san francisco']]):\n",
+    "        return \"It's sunny in San Francisco, but you better look out if you're a Gemini 😈.\"\n",
+    "    else:\n",
+    "        return f\"I am not sure what the weather is in {location}\"\n",
+    "\n",
+    "tools = [get_weather]\n",
+    "\n",
+    "model = model.bind_tools(tools)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Define nodes and edges\n",
+    "\n",
+    "Next let's define our nodes and edges. In our basic ReAct agent there are only two nodes, one for calling the model and one for using tools, however you can modify this basic structure to work better for your use case. The tool node we define here is a simplified version of the prebuilt [`ToolNode`](https://langchain-ai.github.io/langgraph/how-tos/tool-calling/), which has some additional features.\n",
+    "\n",
+    "Perhaps you want to add a node for [adding structured output](https://langchain-ai.github.io/langgraph/how-tos/react-agent-structured-output/) or a node for executing some external action (sending an email, adding a calendar event, etc.). Maybe you just want to change the way the `call_model` node works and how `should_continue` decides whether to call tools - the possibilities are endless and LangGraph makes it easy to customize this basic structure for your specific use case."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import json\n",
+    "from langchain_core.messages import ToolMessage, SystemMessage\n",
+    "from langchain_core.runnables import RunnableConfig\n",
+    "\n",
+    "tools_by_name = {tool.name: tool for tool in tools}\n",
+    "# Define our tool node\n",
+    "def tool_node(state: AgentState):\n",
+    "    outputs = []\n",
+    "    for tool_call in state['messages'][-1].tool_calls:\n",
+    "        tool_result = tools_by_name[tool_call[\"name\"]].invoke(\n",
+    "            tool_call[\"args\"]\n",
+    "        )\n",
+    "        outputs.append(\n",
+    "            ToolMessage(\n",
+    "                content=json.dumps(tool_result),\n",
+    "                name=tool_call[\"name\"],\n",
+    "                tool_call_id=tool_call[\"id\"],\n",
+    "            )\n",
+    "        )\n",
+    "    return {\"messages\": outputs}\n",
+    "\n",
+    "# Define the node that calls the model\n",
+    "def call_model(\n",
+    "    state: AgentState,\n",
+    "    config: RunnableConfig,\n",
+    "):\n",
+    "    system_prompt = SystemMessage(\"You are a helpful AI assistant, please respond to the users query to the best of your ability!\")\n",
```

Comment:
let's add a note that this is similar to customizing the `create_react_agent` with `state_modifier`, but a lot more flexible